File: fm-domain/src/main/java/io/zeitwert/fm/building/model/base/ObjBuildingPartRatingBase.java
package io.zeitwert.fm.building.model.base;

import static io.dddrive.util.Invariant.requireThis;

import java.time.LocalDate;

import io.dddrive.core.ddd.model.Part;
import io.dddrive.core.ddd.model.PartRepository;
import io.dddrive.core.obj.model.base.ObjPartBase;
import io.dddrive.core.oe.model.ObjUser;
import io.dddrive.core.property.model.BaseProperty;
import io.dddrive.core.property.model.EnumProperty;
import io.dddrive.core.property.model.PartListProperty;
import io.dddrive.core.property.model.Property;
import io.dddrive.core.property.model.ReferenceProperty;
import io.zeitwert.fm.building.model.ObjBuilding;
import io.zeitwert.fm.building.model.ObjBuildingPartElementRating;
import io.zeitwert.fm.building.model.ObjBuildingPartRating;
import io.zeitwert.fm.building.model.enums.CodeBuildingMaintenanceStrategy;
import io.zeitwert.fm.building.model.enums.CodeBuildingPart;
import io.zeitwert.fm.building.model.enums.CodeBuildingPartCatalog;
import io.zeitwert.fm.building.model.enums.CodeBuildingRatingStatus;
import kotlin.Pair;

public abstract class ObjBuildingPartRatingBase extends ObjPartBase<ObjBuilding>
		implements ObjBuildingPartRating {

	//@formatter:off
	protected final EnumProperty<CodeBuildingPartCatalog> partCatalog= this.addEnumProperty("partCatalog", CodeBuildingPartCatalog.class);
	protected final EnumProperty<CodeBuildingMaintenanceStrategy> maintenanceStrategy=this.addEnumProperty("maintenanceStrategy", CodeBuildingMaintenanceStrategy.class);
	protected final EnumProperty<CodeBuildingRatingStatus> ratingStatus= this.addEnumProperty("ratingStatus", CodeBuildingRatingStatus.class);
	protected final BaseProperty<LocalDate> ratingDate= this.addBaseProperty("ratingDate", LocalDate.class);
	protected final ReferenceProperty<ObjUser> ratingUser= this.addReferenceProperty("ratingUser", ObjUser.class);
	protected final PartListProperty<ObjBuildingPartElementRating> elementList= this.addPartListProperty("elementList", ObjBuildingPartElementRating.class);
	//@formatter:on

	protected Integer elementWeights = null;

	protected ObjBuildingPartRatingBase(ObjBuilding obj, PartRepository<ObjBuilding, ? extends Part<ObjBuilding>> repository, Property<?> property, Integer id) {
		super(obj, repository, property, id);
	}

	@Override
	public void doAfterCreate() {
		super.doAfterCreate();
		this.setRatingStatus(CodeBuildingRatingStatus.OPEN);
	}

	@Override
	public Part<?> doAddPart(Property<?> property, Integer partId) {
		if (property == this.elementList) {
			PartRepository<ObjBuilding, ?> partRepo = getDirectory().getPartRepository(ObjBuildingPartElementRating.class);
			return partRepo.create(getAggregate(), property, partId);
		}
		return super.doAddPart(property, partId);
	}

	@Override
	public void doAfterAdd(Property<?> property, Part<?> part) {
		if (property == this.elementList) {
			if (this.getRatingDate() != null) {
				Integer year = this.getRatingDate().getYear();
				this.getElement(this.getElementCount() - 1).setRatingYear(year);
			}
		}
	}

	@Override
	public void doAfterSet(Property<?> property) {
		if (property == this.partCatalog) {
			this.elementList.clearParts();
			if (this.getPartCatalog() != null) {
				CodeBuildingPartCatalog catalog = this.getPartCatalog();
				for (Pair<CodeBuildingPart, Integer> part : catalog.getParts()) {
					this.addElement(part.getFirst()).setWeight(part.getSecond());
				}
			}
		} else if (property == this.ratingDate) {
			if (this.getRatingDate() != null) {
				Integer year = this.getRatingDate().getYear();
				for (ObjBuildingPartElementRating element : this.getElementList()) {
					element.setRatingYear(year);
				}
			}
		}
	}

	@Override
	public Integer getRatingYear() {
		return this.getRatingDate() != null ? this.getRatingDate().getYear() : null;
	}

	@Override
	public Integer getElementWeights() {
		return this.elementWeights;
	}

	@Override
	public ObjBuildingPartElementRating getElement(CodeBuildingPart buildingPart) {
		return this.elementList.getParts().stream().filter(p -> p.getBuildingPart() == buildingPart).findFirst()
				.orElse(null);
	}

	@Override
	public ObjBuildingPartElementRating addElement(CodeBuildingPart buildingPart) {
		requireThis(this.getElement(buildingPart) == null,
				"unique element for buildingPart [" + buildingPart.getId() + "]");
		ObjBuildingPartElementRating e = this.elementList.addPart(null);
		e.setBuildingPart(buildingPart);
		return e;
	}

	@Override
	public Integer getCondition() {
		if (this.getRatingYear() == null || this.elementWeights == 0) {
			return null;
		}
		Integer condition = 0;
		for (ObjBuildingPartElementRating element : this.getElementList()) {
			if (element.getWeight() != null && element.getWeight() > 0) {
				condition += element.getWeight() * element.getCondition();
			}
		}
		return condition / this.elementWeights;
	}

	@Override
	public Integer getCondition(Integer year) {
		if (year == null || year < this.getRatingYear() || this.elementWeights == 0) {
			return null;
		}
		Integer condition = 0;
		for (ObjBuildingPartElementRating element : this.getElementList()) {
			if (element.getWeight() != null && element.getWeight() > 0) {
				condition += element.getWeight() * element.getCondition(year);
			}
		}
		return condition / this.elementWeights;
	}

	@Override
	protected void doCalcAll() {
		super.doCalcAll();
		this.doCalcVolatile();
	}

	@Override
	protected void doCalcVolatile() {
		super.doCalcVolatile();
		this.calcElementWeights();
	}

	private void calcElementWeights() {
		this.elementWeights = 0;
		for (ObjBuildingPartElementRating element : this.getElementList()) {
			if (element.getWeight() != null && element.getWeight() > 0) {
				this.elementWeights += element.getWeight();
			}
		}
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/building/model/base/ObjBuildingPartRatingBase.kt
package io.zeitwert.fm.building.model.base

import io.dddrive.ddd.model.Part
import io.dddrive.ddd.model.PartMeta
import io.dddrive.ddd.model.PartRepository
import io.dddrive.obj.model.base.ObjPartBase
import io.dddrive.oe.model.ObjUser
import io.dddrive.property.model.Property
import io.zeitwert.fm.building.model.ObjBuilding
import io.zeitwert.fm.building.model.ObjBuildingPartElementRating
import io.zeitwert.fm.building.model.ObjBuildingPartRating
import io.zeitwert.fm.building.model.enums.CodeBuildingMaintenanceStrategy
import io.zeitwert.fm.building.model.enums.CodeBuildingPart
import io.zeitwert.fm.building.model.enums.CodeBuildingPartCatalog
import io.zeitwert.fm.building.model.enums.CodeBuildingRatingStatus
import java.time.LocalDate

abstract class ObjBuildingPartRatingBase protected constructor(
	obj: ObjBuilding,
	repository: PartRepository<ObjBuilding, out Part<ObjBuilding>>,
	property: Property<*>,
	id: Int,
) : ObjPartBase<ObjBuilding>(obj, repository, property, id),
	ObjBuildingPartRating,
	PartMeta<ObjBuilding> {

	protected val _partCatalog = addEnumProperty("partCatalog", CodeBuildingPartCatalog::class.java)
	protected val _maintenanceStrategy =
		addEnumProperty("maintenanceStrategy", CodeBuildingMaintenanceStrategy::class.java)
	protected val _ratingStatus = addEnumProperty("ratingStatus", CodeBuildingRatingStatus::class.java)
	protected val _ratingDate = addBaseProperty("ratingDate", LocalDate::class.java)
	protected val _ratingUser = addReferenceProperty("ratingUser", ObjUser::class.java)
	protected val _elementList = addPartListProperty("elementList", ObjBuildingPartElementRating::class.java)

	override var elementWeights: Int = 0

	override fun doAfterCreate() {
		super.doAfterCreate()
		this.ratingStatus = CodeBuildingRatingStatus.OPEN
	}

	override fun doAddPart(
		property: Property<*>,
		partId: Int?,
	): Part<*> {
		if (property === this._elementList) {
			val partRepo: PartRepository<ObjBuilding, *> =
				directory.getPartRepository<ObjBuilding, ObjBuildingPartElementRating>(
					ObjBuildingPartElementRating::class.java,
				)
			return partRepo.create(aggregate, property, partId)
		}
		return super.doAddPart(property, partId)
	}

	override fun doAfterAdd(
		property: Property<*>,
		part: Part<*>?,
	) {
		if (property === this._elementList) {
			val ratingDate = this.ratingDate
			if (ratingDate != null) {
				val year: Int? = ratingDate.year
				this.getElement(this.elementCount - 1).ratingYear = year
			}
		}
	}

	override fun doAfterSet(property: Property<*>) {
		if (property === this._partCatalog) {
			// Skip auto-populating elements when loading from persistence
			// (elements will be loaded from DB). Only populate on create/update.
			if (!this.meta.isInLoad) {
				this._elementList.clearParts()
				val partCatalog = this.partCatalog
				if (partCatalog != null) {
					for (part in partCatalog.getParts()) {
						this.addElement(part.first).weight = part.second
					}
				}
			}
		} else if (property === this._ratingDate) {
			val ratingDate = this.ratingDate
			if (ratingDate != null) {
				val year: Int? = ratingDate.year
				for (element in this.elementList) {
					element.ratingYear = year
				}
			}
		}
	}

	override val ratingYear: Int?
		get() = if (this.ratingDate != null) this.ratingDate!!.year else null

	override fun getElement(buildingPart: CodeBuildingPart): ObjBuildingPartElementRating =
		this._elementList.parts
			.stream()
			.filter { p: ObjBuildingPartElementRating? -> p!!.buildingPart === buildingPart }
			.findFirst()
			.orElse(null)

	override fun addElement(buildingPart: CodeBuildingPart): ObjBuildingPartElementRating {
// 		requireThis(this.getElement(buildingPart) == null, "unique element for buildingPart [" + buildingPart.getId() + "]");
		val e: ObjBuildingPartElementRating = this._elementList.addPart(null)
		e.buildingPart = buildingPart
		return e
	}

	override val condition: Int?
		get() {
			if (this.ratingYear == null || this.elementWeights == 0) {
				return null
			}
			var condition = 0
			for (element in this.elementList) {
				if (element.weight != null && element.weight!! > 0) {
					condition += element.weight!! * element.condition!!
				}
			}
			return condition / this.elementWeights!!
		}

	override fun getCondition(year: Int): Int? {
		if (year < this.ratingYear!! || this.elementWeights == 0) {
			return null
		}
		var condition = 0
		for (element in this.elementList) {
			if (element.weight != null && element.weight!! > 0) {
				condition += element.weight!! * element.getCondition(year)
			}
		}
		return condition / this.elementWeights!!
	}

	override fun doCalcAll() {
		super.doCalcAll()
		this.doCalcVolatile()
	}

	override fun doCalcVolatile() {
		super.doCalcVolatile()
		this.calcElementWeights()
	}

	private fun calcElementWeights() {
		var elementWeights = 0
		for (element in this.elementList) {
			if (element.weight != null && element.weight!! > 0) {
				elementWeights += element.weight!!
			}
		}
		this.elementWeights = elementWeights
	}

}
